Userland/Libraries/LibGfx/WebPLoader.cpp

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright (c) 2023, Nico Weber <thakis@chromium.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Debug.h>
+#include <AK/Endian.h>
+#include <AK/Format.h>
+#include <LibGfx/WebPLoader.h>
+
+// Container: https://developers.google.com/speed/webp/docs/riff_container
+// Lossless format: https://developers.google.com/speed/webp/docs/webp_lossless_bitstream_specification
+// Lossy format: https://datatracker.ietf.org/doc/html/rfc6386
+
+namespace Gfx {
+
+namespace {
+
+struct FourCC {
+    constexpr FourCC(char const* name)
+    {
+        cc[0] = name[0];
+        cc[1] = name[1];
+        cc[2] = name[2];
+        cc[3] = name[3];
+    }
+
+    bool operator==(FourCC const&) const = default;
+    bool operator!=(FourCC const&) const = default;
+
+    char cc[4];
+};
+
+// https://developers.google.com/speed/webp/docs/riff_container#webp_file_header
+struct WebPFileHeader {
+    FourCC riff;
+    LittleEndian<u32> file_size;
+    FourCC webp;
+};
+static_assert(AssertSize<WebPFileHeader, 12>());
+
+struct ChunkHeader {
+    FourCC chunk_type;
+    LittleEndian<u32> chunk_size;
+};
+static_assert(AssertSize<ChunkHeader, 8>());
+
+struct Chunk {
+    FourCC type;
+    ReadonlyBytes data;
+};
+
+}
+
+struct WebPLoadingContext {
+    enum State {
+        NotDecoded = 0,
+        Error,
+        HeaderDecoded,
+        SizeDecoded,
+        ChunksDecoded,
+        BitmapDecoded,
+    };
+    State state { State::NotDecoded };
+    ReadonlyBytes data;
+
+    RefPtr<Gfx::Bitmap> bitmap;
+
+    Optional<ReadonlyBytes> icc_data;
+};
+
+static ErrorOr<void> decode_webp_header(WebPLoadingContext& context)
+{
+    // https://developers.google.com/speed/webp/docs/riff_container#webp_file_header
```

Comment:
Please put spec links above the function if it applies to the entire body (LibWeb/LibJS style)